### PR TITLE
[Sanitizer] Support only device compilation on CPU device

### PR DESCRIPTION
**Pull request by zhaomaosu**

### CODE REVIEW

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 1]
```diff
@@ -280,6 +286,9 @@ void SanitizerInterceptor::postLaunchKernel(ur_kernel_handle_t Kernel,
             AH->LID0, AH->LID1, AH->LID2, AH->GID0, AH->GID1, AH->GID2);
         context.logger.always("  #0 {} {}:{}", Func, File, AH->Line);
         if (!AH->IsRecover) {
+            if (m_ShadowMemInited && !DestroyShadowMem()) {
+                context.logger.error("Failed to destory shadow memory");
+            }
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:291]: Is this necessary? "exit(1)" will not call destructor?

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 2]
```diff
@@ -101,7 +101,13 @@ std::string getKernelName(ur_kernel_handle_t Kernel) {
 } // namespace
 
 SanitizerInterceptor::SanitizerInterceptor()
-    : m_IsInASanContext(IsInASanContext()) {}
+    : m_IsInASanContext(IsInASanContext()), m_ShadowMemInited(false) {}
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:104]: if m_IsInASanContext = true, then m_ShadowMemInited = true

File: source/loader/layers/sanitizer/asan_interceptor.cpp [hunk 3]
```diff
@@ -289,20 +298,21 @@ ur_result_t SanitizerInterceptor::allocShadowMemory(
...
+            bool Result;
+            std::call_once(OnceFlag, [&]() {
+                Result = SetupShadowMem();
+            });
+
+            if (!Result) {
+                context.logger.error("Failed to allocate shadow memory");
+                return UR_RESULT_ERROR_OUT_OF_RESOURCES;
+            }
+            m_ShadowMemInited = true;
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_interceptor.cpp:311]:
```cpp
if (!m_ShadowMemInited) {
  ...
}
```

Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_interceptor.cpp:311]: How about remove "m_ShadowMemInited", and add a static variable here? This variable is too general, if we use this variable, we also need to use it for PVC and other devices.

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 1 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Removed

File: source/loader/layers/sanitizer/asan_interceptor.cpp (hunk 2 above)
Comment (zhaomaosu) [reply source/loader/layers/sanitizer/asan_interceptor.cpp]: Done

**Review by wenju-he — APPROVED** [review]
LGTM

**Pull request merged**